src/cursor/abstract_cursor.ts

```diff
@@ -0,0 +1,697 @@
+      throw new MongoError(`flag ${flag} must be a boolean value`);
+    }
+
+    this[kOptions][flag] = value;
+    return this;
+  }
+
+  /**
+   * Map all documents using the provided function
+   *
+   * @param transform - The mapping transformation method.
+   */
+  map(transform: (doc: Document) => Document): this {
+    const oldTransform = this[kTransform];
+    if (oldTransform) {
+      this[kTransform] = doc => {
+        return transform(oldTransform(doc));
+      };
+    } else {
+      this[kTransform] = transform;
+    }
+
+    return this;
+  }
+
+  /**
+   * Set the ReadPreference for the cursor.
+   *
+   * @param readPreference - The new read preference for the cursor.
+   */
+  setReadPreference(readPreference: ReadPreferenceLike): this {
+    if (readPreference instanceof ReadPreference) {
+      this[kOptions].readPreference = readPreference;
+    } else if (typeof readPreference === 'string') {
+      this[kOptions].readPreference = ReadPreference.fromString(readPreference);
+    } else {
+      throw new TypeError('Invalid read preference: ' + readPreference);
+    }
+
+    return this;
+  }
+
+  /**
+   * Set a maxTimeMS on the cursor query, allowing for hard timeout limits on queries (Only supported on MongoDB 2.6 or higher)
+   *
+   * @param value - Number of milliseconds to wait before aborting the query.
+   */
+  maxTimeMS(value: number): this {
+    if (typeof value !== 'number') {
+      throw new TypeError('maxTimeMS must be a number');
+    }
+
+    this[kOptions].maxTimeMS = value;
+    return this;
+  }
+
+  /**
+   * Set the batch size for the cursor.
+   *
+   * @param value - The number of documents to return per batch. See {@link https://docs.mongodb.com/manual/reference/command/find/|find command documentation}.
+   */
+  batchSize(value: number): this {
+    if (this[kOptions].tailable) {
+      throw new MongoError('Tailable cursors do not support batchSize');
+    }
+
+    if (typeof value !== 'number') {
+      throw new TypeError('batchSize requires an integer');
+    }
+
+    this[kOptions].batchSize = value;
+    return this;
+  }
+
+  /* @internal */
+  abstract _initialize(
+    session: ClientSession | undefined,
+    callback: Callback<ExecutionResult>
+  ): void;
+
+  /* @internal */
+  _getMore(batchSize: number, callback: Callback<Document>): void {
+    const cursorId = this[kId];
+    const cursorNs = this[kNamespace];
+    const server = this[kServer];
+
+    if (cursorId == null) {
+      callback(new MongoError('Unable to iterate cursor with no id'));
+      return;
+    }
+
+    if (server == null) {
+      callback(new MongoError('Unable to iterate cursor without selected server'));
+      return;
+    }
+
+    server.getMore(
+      cursorNs.toString(),
+      cursorId,
+      {
+        ...this[kOptions],
+        session: this[kSession],
+        batchSize
+      },
+      callback
+    );
+  }
+}
+
+function nextDocument(cursor: AbstractCursor): Document | null | undefined {
```

Comment:
Because we very much don't want people to actually override the implementation. A TypeScript marking will hide it from your IDE, but it won't prevent a user from monkey-patching the behavior. A few months ago we discussed an even more aggressive approach of passing the methods in as options to the `AbstractCursor` (a la Node.js streams) but I figured this was slightly more readable.

Reply:
Discussed offline: It's important to us to maintain a barrier for overriding behavior we want to have guaranteed.